### ISSUE TITLE
mkl-dnn and awd-lstm-lm

### ISSUE THREAD

**Issue by wohlbier** [intel/mkl-dnn#376]
Hi,
I'm trying to determine if mkl-dnn would have an impact on CPU performance for the Salesforce language model toolkit https://github.com/salesforce/awd-lstm-lm

Insights are appreciated. Thanks!
jgw

**Comment by mgouicem** [comment]
Hi @wohlbier ,
It seems that you are using PyTorch. Pytorch has MKL-DNN integrated and most of the functions you use seem to be covered: 
- for QRNN, convolutions, linear and element wise layers are covered by MKL-DNN, 
- for RNN, LSTM and GRU, are covered by MKL-DNN.

You can check this by running your model and setting the environment variable `MKLDNN_VERBOSE=1`.
If you see no MKL-DNN convolutions or LSTM layer running, then it is likely a pytorch issue (@mingfeima can help with that). If you see MKL-DNN is called but performance is not satisfactory, please post the output of the verbose and we will dig into that.

**Comment by wohlbier** [comment]
@mgouicem Thanks for the reply! I am in fact using pytorch. I set the variable MKLDNN_VERBOSE=1 and I do not see any output that suggests it is from mkl-dnn. Notice that in https://github.com/salesforce/awd-lstm-lm they are using https://github.com/salesforce/pytorch-qrnn. I cannot tell if that model should be invoking mkl-dnn.

**Comment by mingfeima** [comment]
@wohlbier wow, small world...
MKLDNN RNN APIs are not integrated into official PyTorch yet, anyway the work is pretty much done on my side and i will PR soon.
MKLDNN RNN only has standard flavors, e.g. LSTM, GRU, RNN_TANH and RNN_RELU. QRNN is not included.
So is QRNN a popular flavor of RNN now or it is only of limited usage?

**Comment by wohlbier** [comment]
@mingfeima QRNN's are gaining in popularity, I'd say.

If I have a LSTM I want to use with MKLDNN, do I build PyTorch from scratch and set USE_MKLDNN=1?

**Comment by vpirogov** [comment]
Closing as the question seem to be addressed.

**Comment by wohlbier** [comment]
@mingfeima is the PR referred to above the best way to test the effect of MKL-DNN in pytorch? I'm working with an LSTM in pytorch-cpu 1.0.1 and it is not clear to me if I should expect performance benefits if I used your PR. Thanks.

**Comment by mingfeima** [comment]
@wohlbier this PR can't be merged in a short while because the approach for caching rnn weights for mkl-dnn is not favored by pytorch facebook engineers.
This issue can be resolved by a new tensor type design, however i am struggling in pushing my new design and not sure if i can pursue.
Anyway, the performance and functionality are both tested locally. for a typical inference case, the speedup on Xeon could be ~60%, this may vary depend on your RNN input, `input size`, `hidden size`, etc. For training it may not be valid, reason is training with dropout will not call mkl-dnn but vanilla cpu rnn.

**Comment by wohlbier** [comment]
@mingfeima sorry to comment on this old issue, but I can't find the answer to this question anywhere. Do you know how the conda distributions of pytorch-cpu interact with OpenMP runtime libraries? I have seen evidence that both libgomp and libiomp are linked into these builds, but I can't figure out if I can control which runtime library gets called. Thanks for the help.

**Comment by vpirogov** [comment]
@wohlbier, you are more likely to get help if you ask this question on Pytorch github.

**Comment by wohlbier** [comment]
Thanks, I have done so.

**Comment by mingfeima** [comment]
@wohlbier this is an old legacy issue:
the thing is in old versions of PyTorch (probably a year ago), it did link both `iomp` and `gomp` at the same time (`iomp` is loaded from `mklml` which is subset of `mkl`). `iomp5` is actually used since it is loaded prior to `gomp`. when pytorch and caffe2 merges, FB requested that only `gomp` should be in use, so we modified the way `mkl-dnn` building along with pytorch.

so to sum up, the latest version should only link `gomp`.
that is result on my machine:
```bash
(pytorch-mingfei) [mingfeim@mlt-skx086 lib]$ ldd libtorch.so
        linux-vdso.so.1 =>  (0x00007ffc6b363000)
        libgomp.so.1 => /home/mingfeim/anaconda3/envs/pytorch-mingfei/lib/libgomp.so.1 (0x00007fde6a963000)
        libpthread.so.0 => /usr/lib64/libpthread.so.0 (0x00007fde6a733000)
        libmpi.so.20 => /home/mingfeim/anaconda3/envs/pytorch-mingfei/lib/libmpi.so.20 (0x00007fde6a457000)
        librt.so.1 => /usr/lib64/librt.so.1 (0x00007fde6a24f000)
        libgcc_s.so.1 => /home/mingfeim/anaconda3/envs/pytorch-mingfei/lib/libgcc_s.so.1 (0x00007fde6a23a000)
        libdl.so.2 => /usr/lib64/libdl.so.2 (0x00007fde6a036000)
        libmkl_intel_lp64.so => /home/mingfeim/anaconda3/envs/pytorch-mingfei/lib/libmkl_intel_lp64.so (0x00007fde694be000)
        libmkl_gnu_thread.so => /home/mingfeim/anaconda3/envs/pytorch-mingfei/lib/libmkl_gnu_thread.so (0x00007fde67c6a000)
        libmkl_core.so => /home/mingfeim/anaconda3/envs/pytorch-mingfei/lib/libmkl_core.so (0x00007fde63995000)
        libm.so.6 => /usr/lib64/libm.so.6 (0x00007fde63693000)
        libirc.so => /opt/intel/lib/intel64/libirc.so (0x00007fde63425000)
        libc10.so => /home/mingfeim/pytorch/pytorch-mingfei/build/lib/libc10.so (0x00007fde631e6000)
        libstdc++.so.6 => /home/mingfeim/anaconda3/envs/pytorch-mingfei/lib/libstdc++.so.6 (0x00007fde630a5000)
        libc.so.6 => /usr/lib64/libc.so.6 (0x00007fde62ce1000)
        /lib64/ld-linux-x86-64.so.2 (0x0000562f6942f000)
        libopen-rte.so.20 => /home/mingfeim/anaconda3/envs/pytorch-mingfei/lib/./libopen-rte.so.20 (0x00007fde62a60000)
        libopen-pal.so.20 => /home/mingfeim/anaconda3/envs/pytorch-mingfei/lib/./libopen-pal.so.20 (0x00007fde6276e000)
        libutil.so.1 => /usr/lib64/libutil.so.1 (0x00007fde6256b000)
```

if you are using conda distribution, find the linkage of pytorch by:
```bash
python -c 'import torch; print(torch.__path__)'
['/home/mingfeim/anaconda3/envs/pytorch-mingfei/lib/python3.7/site-packages/torch']
ldd /home/mingfeim/anaconda3/envs/pytorch-mingfei/lib/python3.7/site-packages/torch/lib/libtorch.so
```

if you find it is not linked only `gomp`, report an issue to PyTorch github, we will take a look at that.

---
Aside from this topoic: Sometimes, one may still want to use `iomp5`, you can use the `PRE_LOAD` to do this, [ref](https://gist.github.com/mingfeima/363a9ab850be54d5837f9cc542ad2b38#4-use-intel-openmp-library)

 And reference on how to set environment variables on `gomp` or `iomp`: [ref](https://software.intel.com/en-us/articles/how-to-get-better-performance-on-pytorchcaffe2-with-intel-acceleration)
